Changed kubectl config set-cluster and set-credentials to support process substitution for filenames

brianpursley opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
`kubectl config set-cluster` doesn't work with process substitution, so you can't set the certificate authority from an environment variable like this:
```
kubectl config set-cluster xyz --embed-certs --certificate-authority <(echo $CACERT)
```

This doesn't result in an error, but fails to set the certificate authority.

This PR changes `kubectl config set-cluster` so that it only reads the ca file once, thereby allowing process substitution to work properly.

This PR also makes the same change to `kubectl config set-credentials`.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubectl/issues/501

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs

```

brianpursley commented:
/cc @liggitt 

This new PR replaces https://github.com/kubernetes/kubernetes/pull/90068 and takes a different approach, fixing the underlying problem preventing the use of fifos with the existing --certificate-authority flag

I would like to write a test to prevent this from regressing in the future, but given the `ioutil.ReadFile` dependency, I'm not sure where the best place would be to do that. 

I was considering creating an interface for reading the file and use that instead of ioutil.ReadFile directly, so that it can be mocked and make sure it is only ever read once.

I don't want to overkill this.

Any thoughts on whether that is a good idea, or if there is a simpler way to unit test this?

brianpursley commented:
/cc @NathanHowell

brianpursley commented:
/retest

brianpursley commented:
/retest

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/config/create_cluster.go:
```diff
@@ -108,8 +108,10 @@ func (o createClusterOptions) run() error {
 	}
-	cluster := o.modifyCluster(*startingStanza)
-	config.Clusters[o.name] = &cluster
+	config.Clusters[o.name], err = o.modifyCluster(*startingStanza)
+	if err != nil {
+		return err
+	}
```
```suggestion
	cluster, err := o.modifyCluster(*startingStanza)
	if err != nil {
		return err
	}
	config.Clusters[o.name] = &cluster
```

brianpursley replied:
Made the change with a slight modification, because modifyCluster return type has changed from `clientcmdapi.Cluster` to `*clientcmdapi.Cluster` so that nil can be returned in case of an error.

```
config.Clusters[o.name] = cluster
```

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/config/create_cluster.go:
```diff
@@ -181,9 +187,6 @@ func (o createClusterOptions) validate() error {
 		if caPath == "" {
 			return fmt.Errorf("you must specify a --%s to embed", clientcmd.FlagCAFile)
 		}
-		if _, err := ioutil.ReadFile(caPath); err != nil {
```
can we os.Stat here to validate it exists?

brianpursley replied:
Yes, that appears to work fine, even with a fifo... I'll make that change

brianpursley replied:
Good idea!

liggitt commented:
add a test to test/cmd/kubeconfig.sh that exercises redirection, similar to:

https://github.com/kubernetes/kubernetes/blob/438debfb7263619d444c2eb9f6b52747415a9f24/test/cmd/core.sh#L846-L849

liggitt commented:
it looks like create_authinfo has a similar issue with the client cert/key args... mind making the same change in that file and adding a test for that command as well?

brianpursley commented:
Sorry... git issue.  Will fix it.

brianpursley commented:
@liggitt OK, its ready for another look. Just changed ReadFile to Stat instead and added tests for `kubectl config set-cluster` and `kubectl config set-credentials`

liggitt commented:
/retest

liggitt commented:
some shellcheck issues with the test (see https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/91077/pull-kubernetes-verify/1261050358264762370)

brianpursley commented:
> some shellcheck issues with the test (see https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/91077/pull-kubernetes-verify/1261050358264762370)

Thanks... I didn’t even know about spellcheck.  I’ll take care of it.

zhouya0 commented:
/retest

brianpursley committed:
Changed kubectl config set-cluster and set-credentials to support process substitution for filenames

brianpursley commented:
/retest

brianpursley commented:
/assign @liggitt

liggitt commented:
/lgtm
/approve
